auth/impersonate/impersonate.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package impersonate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"cloud.google.com/go/auth"
+	"cloud.google.com/go/auth/detect"
+	"cloud.google.com/go/auth/httptransport"
+	"cloud.google.com/go/auth/internal"
+)
+
+var (
+	iamCredentialsEndpoint = "https://iamcredentials.googleapis.com"
+	oauth2Endpoint         = "https://oauth2.googleapis.com"
+)
+
+// NewCredentialTokenProvider returns an impersonated
+// [cloud.google.com/go/auth/TokenProvider] configured with the provided options
+// and using credentials loaded from Application Default Credentials as the base
+// credentials if not provided with the opts.
```

Comment:
I think it might be helpful to link to `NewIDTokenProvider` here and explain the difference between `NewIDTokenProvider` and `NewCredentialTokenProvider` somehow.

Reply:
Will mention in doc.go